File: py/advent_of_code_2020/day02.py
import sys
import re


def read_policy_and_password(input):
    x = re.match(r"(\d+)-(\d+) (\w)\: (\w+)", input)
    return (int(x[1]), int(x[2]), x[3], x[4])


def validate_password_against_policy(password, letter, lower, upper):
    n = len([c for c in password if c == letter])
    return lower <= n <= upper


def validate_policy_and_password(input):
    lower, upper, letter, password = read_policy_and_password(input)
    return validate_password_against_policy(password, letter, lower, upper)


def validate_password_against_policy2(password, letter, p1, p2):
    return (p1 <= len(password) and password[p1-1] == letter) ^ \
           (p2 <= len(password) and password[p2-1] == letter)


def validate_policy_and_password2(input):
    p1, p2, letter, password = read_policy_and_password(input)
    return validate_password_against_policy2(password, letter, p1, p2)


if __name__ == "__main__":
    with open(sys.argv[1], "r") as fp:
        input = [line.rstrip() for line in fp.readlines()]
    print("Part 1:")
    n = len([x for x in input if validate_policy_and_password(x)])
    print(n)
    print("Part 2:")
    n = len([x for x in input if validate_policy_and_password2(x)])
    print(n)

File: scala/src/main/scala/day02/Day02.scala
package day02

import scala.io.Source
import scala.util.matching.Regex

case class PolicyValidation(p1: Int, p2: Int, letter: Char, password: String) {

  def validate() = {
    val occurrences = password.filter(_ == letter).length
    p1 <= occurrences && occurrences <= p2
  }

  def validate2() = {
    val x1 = p1 <= password.length && password(p1-1) == letter
    val x2 = p2 <= password.length && password(p2-1) == letter
    x1 != x2
  }
}

object Day02 {

  def readPolicyAndPassword(input: String): PolicyValidation = {
    val info = raw"(\d+)-(\d+) (\w)\: (\w+)".r
    input match {
      case info(lower, upper, letter, password) =>
        new PolicyValidation(lower.toInt, upper.toInt, letter(0), password)
    }
  }

  def main(args: Array[String]): Unit = {
    val input: Vector[PolicyValidation] =
      Source.fromFile(args(0)).getLines.toVector.map(readPolicyAndPassword(_))
    println("Part 1:")
    println(input.filter(_.validate).length)
    println("Part 2:")
    println(input.filter(_.validate2).length)
  }
}
